rust/src/pgsql/pgsql.rs

```diff
@@ -380,10 +409,31 @@ impl PgsqlState {
                     // https://samadhiweb.com/blog/2013.04.28.graphviz.postgresv3.html
                     if let Some(tx) = self.find_or_create_tx() {
                         tx.tx_data.updated_ts = true;
-                        tx.request = Some(request);
                         if let Some(state) = new_state {
-                            if Self::request_is_complete(state) {
-                                // The request is always complete at this point
+                            if state == PgsqlStateProgress::FirstCopyDataInReceived
+                            || state == PgsqlStateProgress::ConsolidatingCopyDataIn {
+                                // here we're actually only counting how many messages were received.
+                                // frontends are not forced to send one row per message
+                                if let PgsqlFEMessage::ConsolidatedCopyDataIn(msg) = request {
+                                    tx.sum_data_size(msg.data_size);
+                                    tx.incr_row_cnt();
+                                }
+                            } else if state == PgsqlStateProgress::CopyDoneReceived && tx.get_row_cnt() > 0 {
+                                let dummy_request = PgsqlFEMessage::ConsolidatedCopyDataIn(
+                                    ConsolidatedDataRowPacket {
+                                        identifier: b'd',
+                                        row_cnt: tx.get_row_cnt(),
+                                        data_size: tx.data_size, // total byte count of all copy_data messages combined
+                                    },
+                                );
+                                tx.requests.push(dummy_request);
+                                tx.requests.push(request);
+                                // rest values
+                                tx.data_row_cnt = 0;
+                                tx.data_size = 0;
```

Comment:
These don't really seem necessary, from what I can think, but they got merged for the CopyOut mode, so thought I'd keep the pattern

Reply:
Why do we add a "dummy" request ?

Reply:
The `dummy_request` is the consolidated data of the DataRows. Probably a bad variable name... 🤔

Reply:
Could it not be all in a single request object ? (would it make more sense ?)

Reply:
We keep all consolidated DataRows or copy data in a single requests object, in the end. But there will be a next request message that is the `CopyDone`, and thus we have two requests by the end of this side of the transaction. 
If we abstracted the `CopyDone` message, then I suppose we could, but would be a different pattern than what we're doing for when we parse similar messages from the server/ backend side.

For now, I'm going with renaming these variables, as it was a bad name choice.